lib/Dialect/SV/Transforms/HWMemSimImpl.cpp

```diff
@@ -478,14 +482,23 @@ void HWMemSimImpl::generateMemory(HWModuleOp op, FirMemory mem) {
           b.getStringAttr(moduleNamespace.newName(reg.getName())));
 
     if (mem.initIsInline) {
-      b.create<sv::IfDefOp>("SYNTHESIS", std::function<void()>(), [&]() {
+      if (initializeMemSynthesis) {
         b.create<sv::InitialOp>([&]() {
           b.create<sv::ReadMemOp>(reg, mem.initFilename,
                                   mem.initIsBinary
                                       ? MemBaseTypeAttr::MemBaseBin
                                       : MemBaseTypeAttr::MemBaseHex);
         });
-      });
+      } else {
+        b.create<sv::IfDefOp>("SYNTHESIS", std::function<void()>(), [&]() {
+          b.create<sv::InitialOp>([&]() {
+            b.create<sv::ReadMemOp>(reg, mem.initFilename,
+                                    mem.initIsBinary
+                                        ? MemBaseTypeAttr::MemBaseBin
+                                        : MemBaseTypeAttr::MemBaseHex);
+          });
```

Comment:
This code is duplicated from above. It's too low level to put in a static function, but a lambda would avoid duplication.

Reply:
Consolidated into a lambda. Thanks for the tip... I didn't knew exactly the best solution to this but knew it was ugly.